### PR TITLE
More server socket options

### PR DESCRIPTION
**Pull request by mdavidsaver**
Support binding multiple sockets, which can be a mixture of tcp and unix named sockets.  Also write an additional "info" file with pid and a list of server endpoints.

The motivation for these changes is to remove the need to explicitly allocation tcp port numbers.  In fact, it is possible to use only unix sockets in combination with conserver.  In addition, specifying port 0 binds a random port, which can be found in the new "info" file.

Server endpoints can now be specified in one of four ways.
- `<port>` (interface address is determined by the restrict setting as previously)
- `<interface>:<port>`
- `unix:</path/to/sock>`
- `unix:<user>:<group>:<perm>:</path/to/sock>`

The final form allows, for example, access to the socket to be restricted to a certain unix group.  The user, group, and/or permissions may be omitted, so  `-S unix:::/path/sock` can be abbreviate as  `-S unix:/path/sock`.

The `-S/--port <portspec>` argument is introduced, and can be given multiple times, to give additional endpoints to bind.

The `-I,--info-file <path>` argument specifies the location where a file is written containing the procServ PID and a list of all endpoints.  For example `-I /run/procserv-foo/info` might write.

```
pid:12345
tcp:127.0.0.1:4567
unix:/run/procserv-foo/control
```

One change to existing behavior is to delete the PID file on exit.  In addition the new info file, and any unix sockets, are also removed.

An example of using these features can be found [here](https://github.com/mdavidsaver/procServ/tree/master) in the context of work on a systemd replacement for the existing [sysv-rc-softioc](https://github.com/epicsdeb/sysv-rc-softioc) init scripts.

### CODE REVIEW

**Comment by ralphlange** [comment]
That looks very interesting.
I am already past -rc2 on version 2.6.1 (and people are waiting for it), but this looks like a 2.7.0 will be following rather sooner than later.

**Comment by mdavidsaver** [comment]
> 2.7.0

This was my thinking as well.  I'd like to include the systemd support scripts, which aren't done yet.

**Comment by mdavidsaver** [comment]
Add magic logfile ```-L <stdout>``` which can be used with ```-f``` to activate systemd style daemon logging to stdout while ignoring stdin.

Also, skip tty operations on stdin when it isn't a tty.

With this I think the code changes to allow good systemd integration are complete.

**Comment by NickeZ** [comment]
I would use '-' to log to stdout like the rest of the gnu tools and I think logging to stdout shouldn't make "foreground" or "interactive mode" non-interactive.

**Comment by NickeZ** [comment]
I had a theory that you don't even need to change the code to output to stdout, but it didn't work:

procServ[5468]: /usr/bin/procServ: unable to open log file /dev/stdout

**Comment by mdavidsaver** [comment]
> I would use '-' to log to stdout like the rest of the gnu tools

Fair point.  I can't remember why I had ```"<stdout>"``` but it seems reasonable to simply replace with ```"-"```.

**Comment by mdavidsaver** [comment]
> I had a theory that you don't even need to change the code to output to stdout, but it didn't work:

Be aware that there is some baked in trickiness about the relationship between stdin and stdout.

```c++
AddConnection(clientFactory(0));
```

So interactive mode works by *writing* to stdin.  Which works just fine if it has been open()'d as rw, which is the case if it is a tty.

**Comment by NickeZ** [comment]
I didn't look very carefully at the code yet. I would probably have a flag (-I / --interactive) to enable "stdin from terminal to stdin of process" and a separate flag (-f / --foreground) to enable "stdout of process to stdout of terminal". Obviously change the meaning of -f, which might be unacceptable.

This would be closer to what I consider expected behavior [1]. If I start a daemon in the "foreground" I wouldn't expect it to become interactive.

[1]: https://en.wikipedia.org/wiki/Principle_of_least_astonishment

**Comment by mdavidsaver** [comment]
@NickeZ It might be worthwhile to talk about use cases before undertaking this work (hint, I tried to dis-entangle stdin/stdout but decided it was too much trouble).  In any event I think it is outside the scope of my proposed change.  (You might open an issue to discuss this)  My goal is to add the ability for procServ to behave as a systemd "new-style" daemon, and allow management of multiple instances w/o manual allocation of tcp port numbers, without breaking existing (classic? retro?) daemon behavior.

**Comment by mdavidsaver** [comment]
@ralphlange ping

**Comment by mdavidsaver** [comment]
@ralphlange Ping

**Comment by mdavidsaver** [comment]
Added a NULL test missed when changing '<stdin>' to '-'.

Also, add support for linux specific abstract unix sockets.  Abstract unix sockets are a linux specific feature whereby the socket has a "path" name associated with it, but no presence in the filesystem (and no associated permission check).  Analogous to IP socket bound to localhost with a name string instead of a port number.  Denote an abstract socket with a leading '@' before the path (seen in the output of 'netstat').

**Comment by mdavidsaver** [comment]
@ralphlange PIng

**Comment by ralphlange** [comment]
Have you tested the domain sockets on cygwin?

File: acceptFactory.cc [hunk 1]
```diff
@@ -102,10 +178,162 @@ void acceptItem::remakeConnection()
...
+     */
+    abstract = spec[0]=='@';
+#ifndef __linux__
+    if(abstract) {
+        fprintf(stderr, "Abstract unix sockets not supported by this host\n");
+        exit(1);
+    }
+#endif
+
+    memset(&addr, 0, sizeof(0));
```
Comment (ralphlange) [inline acceptFactory.cc]: That looks wrong. Shouldn't it be

```C++
memset(&addr, 0, sizeof(addr));
```

File: acceptFactory.cc [hunk 2]
```diff
@@ -102,10 +178,162 @@ void acceptItem::remakeConnection()
...
+        }
+    }
+
+    _fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (_fd < 0) {
+        PRINTF("Socket error: %s\n", strerror(errno));
+        throw errno;
+    }
+
+    bindStatus = bind(_fd, (struct sockaddr *) &addr, sizeof(addr));
```
Comment (ralphlange) [inline acceptFactory.cc]: In the case of abstract sockets, this is supplying all trailing zeroes to `bind()` as the socket name. (Use `cat /proc/net/unix` to show this.) Better do something like:

```C++
#include <cstddef>

    if (abstract)
        bindStatus = bind(_fd, (struct sockaddr *) &addr, offsetof(struct sockaddr_un, sun_path)
                          + 1 /*\0*/ + strlen(addr.sun_path+1));
    else
        bindStatus = bind(_fd, (struct sockaddr *) &addr, sizeof(addr));
```

**Comment by mdavidsaver** [comment]
> Have you tested the domain sockets on cygwin?

No, and I have no plans to do so.

File: acceptFactory.cc (hunk 2 above)
Comment (mdavidsaver) [inline acceptFactory.cc]: ha, right.  From the manpage

> Null  bytes in the name have no special significance.

Looks like in both cases the length could be set by strlen().  I'll try this.

**Comment by ralphlange** [comment]
> > Have you tested the domain sockets on cygwin?
> No, and I have no plans to do so.

OK, I can do some shallow testing of the sockets on cygwin, and will try to interest Mark (as he's the cygwin power user).

**Comment by ralphlange** [comment]
I can push my changes to the PR tomorrow, once you tried if `strlen()` works in both cases for the call to `bind()`.

**Comment by mdavidsaver** [comment]
Should now be fixed.

**Pull request merged**